pkg/tfbridge/defaults/tokens.go

```diff
@@ -0,0 +1,928 @@
+		if ignored[key] {
+			return true
+		}
+		return each(key, value)
+	})
+}
+
+func sharedPrefix(s1, s2 string) string {
+	// Shorten the longer string so it is only as long as the shortest string
+	if len(s1) < len(s2) {
+		s2 = s2[:len(s1)]
+	} else if len(s1) > len(s2) {
+		s1 = s1[:len(s2)]
+	}
+
+	for i := range s1 {
+		if s1[i] != s2[i] {
+			return s1[:i]
+		}
+	}
+	return s1
+}
+
+type tokenInfo struct{ mod, name string }
+
+func tokenFromMap[T b.ResourceInfo | b.DataSourceInfo](
+	tokenMap map[string]tokenInfo, finalize MakeToken, new func(tk string) *T,
+) ElementStrategy[T] {
+	return func(tfToken string) (*T, error) {
+		info, ok := tokenMap[tfToken]
+		if !ok {
+			existing := []string{}
+			for k := range tokenMap {
+				existing = append(existing, k)
+			}
+			return nil, fmt.Errorf("TF token '%s' not present when prefix computed, found %#v", tfToken, existing)
+		}
+		tk, err := finalize(camelCase(info.mod), upperCamelCase(info.name))
+		if err != nil {
+			return nil, err
+		}
+		return new(tk), nil
+	}
+}
+
+type tokenHistory[T ~string] struct {
+	Current T          `json:"current"`        // the current Pulumi token for the resource
+	Past    []alias[T] `json:"past,omitempty"` // Previous tokens
+
+	MajorVersion int                      `json:"majorVersion,omitempty"`
+	Fields       map[string]*fieldHistory `json:"fields,omitempty"`
+}
+
+type alias[T ~string] struct {
+	Name         T    `json:"name"`         // The previous token.
+	InCodegen    bool `json:"inCodegen"`    // If the alias is a fully generated resource, or just a schema alias.
+	MajorVersion int  `json:"majorVersion"` // The provider's major version when Name was introduced.
+}
+
+type aliasHistory struct {
+	Resources   map[string]*tokenHistory[tokens.Type]         `json:"resources,omitempty"`
+	DataSources map[string]*tokenHistory[tokens.ModuleMember] `json:"datasources,omitempty"`
+}
+
+type fieldHistory struct {
+	MaxItemsOne *bool `json:"maxItemsOne,omitempty"`
+
+	Fields map[string]*fieldHistory `json:"fields,omitempty"`
+	Elem   *fieldHistory            `json:"elem,omitempty"`
+}
+
+// Apply resource and function aliases to maintain backwards compatibility with previous
+// versions of this provider.  Apply MaxItemsOne history to maintain backwards
+// compatibility with previous versions of this provider.
+//
+// All history is dropped when there is a major version change.
+//
+// Panics if ApplyAliases would return an error.
+func MustApplyAliases(providerInfo *b.ProviderInfo)  {
+	err := ApplyAliases(providerInfo)
+	contract.AssertNoErrorf(err, "Failed to apply aliases")
+}
+
+// Apply resource and function aliases to maintain backwards compatibility with previous
+// versions of this provider.  Apply MaxItemsOne history to maintain backwards
+// compatibility with previous versions of this provider.
+//
+// All history is dropped when there is a major version change.
+func ApplyAliases(providerInfo *b.ProviderInfo) error {
```

Comment:
Now there's a ton of subtlety going on here and I'd take the time to write it out.

```
Automatically applies backwards compatibility best practices.

Specifically, ApplyAliases may perform the following actions:

- Call [ResourceInfo.RenameResourceWithAlias] 
- Edit [ResourceInfo.Aliases]
- Edit [SchemaInfo.MaxItemsOne]

The goal is to always maximize backwards compatibility and reduce breaking changes for the users of the Pulumi providers. 

Resource aliases help mask TF provider resource renames or changes in mapped tokens so older programs continue to work.  See [ResourceInfo.RenameResourceWithAlias] and [ResourceInfo.Aliases] for more details.

[SchemaInfo.MaxItemsOne] changes are also important because they involve flattening and pluralizing names. Collections (lists or sets) marked with MaxItems=1 are projected as scalar types in Pulumi SDKs. Therefore changes to the MaxItems property may be breaking the compilation of programs as the type changes from `T to List[T]` or vice versa. To avoid these breaking changes, this method undoes any upstream changes to MaxItems using [SchemaInfo.MaxItemsOne] overrides. This happens until a major version change is detected, and then overrides are cleared. Effectively this makes sure that upstream MaxItems changes are deferred until the next major version.

Implementation note: to operate correctly this method needs to keep a persistent track of a database of past decision history. This is currently done by doing reads and writes to `providerInfo.GetMetadata()`, which is assumed to be persisted across provider releases. The bridge framework keeps this information written out to an opaque `bridge-metadata.json` blob which is expected to be stored in source control to persist across releases.
```

Reply:
The name ApplyAliases is also a bit surprising given that it also applies MaxItemsOne edits. Naming is hard though. ApplyAliesesAndMaxItems is a mouthful. ApplyBackwardsCompatibility is too ambitious. OK with leaving as is given bigger docstring but if you have ideas..